fetcher/fetcher.go

```diff
@@ -16,58 +17,61 @@ import (
 )
 
 // Fetcher retrieves values on-the-fly and ensures that
-// a value is only fetched from disk once. Subsequent
+// a value is only fetched from data once. Subsequent
 // requests can be retrieved from cache.
 type Fetcher struct {
 	im state.Immutable
 
-	keysToFetch map[string]*keyData        // Number of txns waiting for a key
-	txnsToFetch map[ids.ID]*sync.WaitGroup // Number of keys a txn is waiting on
-	fetchLock   sync.RWMutex
+	l    sync.RWMutex
+	keys map[string]*key
+	txs  map[ids.ID]*tx
+	err  error
+	done bool
 
-	stopOnce  sync.Once
-	stop      chan struct{}
-	err       error
-	fetchable chan *task
-	l         sync.RWMutex
-	wg        sync.WaitGroup
+	wg       sync.WaitGroup
+	tasks    chan *task
+	waitOnce sync.Once
 
-	completed int
-	numTxs    int
+	stop     chan struct{}
+	stopOnce sync.Once
```

Comment:
nit: may be cleaner if we change to `doneOnce` and differ the `done` behavior whether we are "complete" or "stopping"

Reply:
Then we can remove this `done` var

Reply:
we can call `finishOnce`